sql/select.go

```diff
@@ -85,3 +99,523 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
 	parser.WalkStmt(&v, stmt)
 	return v.err
 }
+
+// selectIndex analyzes the scanNode to determine if there is an index
+// available that can fulfill the query with a more restrictive scan.
+//
+// The current occurs in two passes. The first pass performs a limited form of
+// value range propagation for the qvalues (i.e. the columns). The second pass
+// takes the value range information and determines which indexes can fulfill
+// the query (we currently only support covering indexes) and selects the
+// "best" index from that set. The cost model based on keys per row, key size
+// and number of index elements used.
+func (p *planner) selectIndex(s *scanNode) (planNode, error) {
+	if s.desc == nil || s.filter == nil {
+		// No table or where-clause.
+		return s, nil
+	}
+
+	rangeInfo := make(qvalueRangeMap)
+	rangeInfo.analyzeExpr(s.filter)
+
+	indexes := make([]*indexInfo, 0, len(s.desc.Indexes)+1)
```

Comment:
sorry, confused the actual index descriptors with the `indexInfo`. This is fine.